Moving to cev2 - publisher, receive adapter - Implementing KRShaped

nachocano opened the pull request:
Fixes #1167
Fixes #1166
Helps with https://github.com/google/knative-gcp/issues/1209

<!-- Please include the 'why' behind your changes if no issue exists -->

## Proposed Changes

- 🧽  Moving publisher and receive adapter to cev2 and using wire.
- 🧽  Updating some of the CE formats due to ongoing conversations. There were undocumented extensions we were adding that are not needed. The release notes of @yolocs PR should include all the breaking changes we are planning to  make.
- 🧽 Implementing KRShaped in all resources otherwise it does not compile

- TODOs
  - Fix adapter UTs to be done in a follow up as it might take sometime and don't want to block @capri-xiyue nor @yolocs  https://github.com/google/knative-gcp/issues/1333
 - Move e2e tests to cev2. https://github.com/google/knative-gcp/issues/1336

<!--
If this change has user-visible impact, follow the instructions below.
Examples include:

- 🎁 Add new feature
- 🐛 Fix bug
- 🧽 Update or clean up current behavior
- 🗑️ Remove feature or internal logic

Otherwise delete the rest of this template.
-->

**Release Note**

<!--
🗒️ If this change has user-visible impact, write a release note in the block
below. Include the string "action required" if additional action is required of
users switching to the new release, for example in case of a breaking change.

Write as if you are speaking to users, not other Knative contributors. If this
change has no user-visible impact, no release-note is needed.
-->

```release-note
We are using CE SDK v2 for Sources and Channels, with all its performance improvements
```

**Docs**

<!--
📖 If this change has user-visible impact, link to an issue or PR in
https://github.com/knative/docs.
-->

nachocano commented:
/assign @Harwayne 
/assign @ian-mi 
/assign @yolocs 

Guys this is a WIP PR, but it is kind of big so I want to get your input asap. Especially on some ugly part of our old receive adapter that IIRC we use for channels. There is a tracing part that I'm not sure what we are actually doing...

@yolocs this will help out with the changes to events schemas. The strings hasn't been changed yet, but I refactored things so that it would be easy to do...

Either way, I'm having seconds thoughts about including this as it's big and kind of last minute. We can always do the changes of the events without moving to cev2, and then do the move...
The upside of doing this now is that I've heard from folks that the old sdk was not very performant. So this might improve the performance of our sources data plane, which we have been neglecting so far.

nachocano commented on pkg/pubsub/adapter/adapter.go:
```diff
@@ -18,253 +18,202 @@ package adapter
+
+		// TODO check if this is OK
 		// Update the tracing information to use the span returned by the transformer.
-		ctx = trace.NewContext(ctx, trace.FromContext(transformedCTX))
-	}
+		// ctx = trace.NewContext(ctx, trace.FromContext(transformedCTX))
+		if span := trace.FromContext(ctx); span.IsRecordingEvents() {
+			span.Annotate(ceclient.EventTraceAttributes(e),
+				"Event reply",
+			)
+		}
```
@ian-mi @Harwayne 
here is the part of tracing that I have no clue about. Is this somewhat equivalent to the old:
`trace.NewContext(ctx, trace.FromContext(transformedCTX))` ?

nachocano replied:
OK, seems that the e2e tests failing are due to tracing problems. I'm surely doing something wrong...

yolocs commented on pkg/pubsub/adapter/adapter.go:
```diff
@@ -18,253 +18,202 @@ package adapter
 
-	if a.reporter == nil {
-		a.reporter = NewStatsReporter()
+	// Done channel has been closed, we need to gracefully shutdown. The cancel() method will start its
+	// shutdown, if it hasn't finished in a reasonable amount of time, just return an error.
+	cancel()
+	select {
+	case err := <-errCh:
+		return err
+	case <-time.After(30 * time.Second):
+		return errors.New("timeout shutting down adapter")
 	}
+}
```
I don't quite understand this logic. To reach `cancel()` it seems that either `Receive` returned an error or context was done. If it's an error (1st case), then it means the pull has already stopped. If the context was done (2nd case), then it should already be handled by the `Receive` function. So it looks to me the `cancel()` is redundant and we just need to wait 30s. Thoughts?

nachocano replied:
I think you are right, I copied/pasted this from some eventing stuff. I'll double check.
Will fix the other stuff first, and then address this.

nachocano replied:
Done, thanks Chen!

yolocs commented on pkg/pubsub/adapter/adapter.go:
```diff
@@ -18,253 +18,202 @@ package adapter
-		transformedCTX, transformedEvent, err := a.transformer.Send(ctx, event)
-		rtctx := cloudevents.HTTPTransportContextFrom(transformedCTX)
+	// TODO consider renaming transformer as it is confusing. And see if we can get rid of this.
+	if a.transformerURI != "" {
+		resp, err := a.sendMsg(ctx, a.transformerURI, message)
```
Are we really hitting this code path?

nachocano replied:
yes, for for replies we are... it's pretty nasty. Added a couple of TODOs in the code.. This has been here forever and we should clean it up

nachocano commented on pkg/pubsub/adapter/adapter.go:
```diff
@@ -18,253 +18,218 @@ package adapter
 		// Update the tracing information to use the span returned by the transformer.
-		ctx = trace.NewContext(ctx, trace.FromContext(transformedCTX))
+		// TODO how to get the span returned by the transformer?
+		ctx = trace.NewContext(ctx, trace.FromContext(ctx))
```
@Harwayne @ian-mi 
I think this is where I'm messing up... Previously we had a context returned and we created a new context with the returned span... not sure what is the proper way of doing this with the new cev2 sdk

nachocano replied:
ok, thanks for the pointer Ian, added some stuff. Let's see if that works. 
Otherwise I'll take a look next week

ian-mi commented on cmd/pubsub/receive_adapter/main.go:
```diff
@@ -72,26 +143,53 @@ func main() {
+		logger.Error("Failed to convert base64 extensions to map: %v", zap.Error(err))
 	}
+
+	ctx = context.WithProjectKey(ctx, projectID)
+	ctx = context.WithTopicKey(ctx, env.Topic)
+	ctx = context.WithSubscriptionKey(ctx, env.Subscription)
```
Can these be adapter args instead of passed through ctx? It might be worthwhile to create an adapter args struct since InitializeAdapter already has so many arguments.

nachocano replied:
yes, let me do that. It's kind of ugly as is right now. Thanks!

nachocano replied:
agreed to add in AdapterArgs only the ones used by the adapter. In this case topic. The others are needed for initializing other objects as well

ian-mi commented on pkg/broker/handler/fanout.go:
```diff
@@ -176,7 +176,6 @@ func (p *FanoutPool) SyncOnce(ctx context.Context) error {
 
 		// Start the handler with broker key in context.
 		hc.Start(handlerctx.WithBrokerKey(ctx, b.Key()), func(err error) {
-			// We will anyway get an error because of https://github.com/cloudevents/sdk-go/issues/470
```
Is this change unrelated?

nachocano replied:
unrelated. Was looking at code and saw that TODO but that bug was fixed so I removed the comment. Let me know if you want it back

nachocano replied:
feel free to unresolved if you want this comment back in

ian-mi commented on pkg/pubsub/adapter/adapter.go:
```diff
@@ -18,253 +18,246 @@ package adapter
+		outbound:               outbound,
+		converter:              converter,
+		reporter:               reporter,
+		sinkURI:                string(sinkURI),
+		transformerURI:         string(transformerURI),
+		namespacedResourceName: types.NamespacedName{Name: string(name), Namespace: string(namespace)},
+		converterType:          converterType,
+		extensions:             extensions,
```
It might simplify the constructor to bundle these into a config struct.

nachocano replied:
Will do an AdapterArgs one and pass it here...

nachocano replied:
done

ian-mi commented on pkg/pubsub/adapter/adapter.go:
```diff
@@ -18,253 +18,246 @@ package adapter
-			return err
+			a.logger.Error("Failed to convert response message to event",
+				zap.Any("response", respMsg), zap.Error(err))
+			msg.Ack()
```
Shouldn't msg be nacked in this case?

nachocano replied:
yes, I think so... thanks for the catch

ian-mi commented on pkg/pubsub/adapter/adapter.go:
```diff
@@ -18,253 +18,246 @@ package adapter
 		// Update the tracing information to use the span returned by the transformer.
-		ctx = trace.NewContext(ctx, trace.FromContext(transformedCTX))
+		var span *trace.Span
+		ctx, span = a.startSpan(ctx, event)
```
I don't think this is right place to start a new span. We should instead start a single span at the start of receive which will be the parent of both the transform and outbound transactions. The span name should probably be derived from the name of the source in this case.

nachocano replied:
updated it. Thanks for the feedback on this one.
Let's see if tests pass

ian-mi commented on pkg/pubsub/adapter/adapter.go:
```diff
@@ -18,253 +18,246 @@ package adapter
+	if resp.StatusCode/100 != 2 {
+		a.logger.Error("Event delivery failed", zap.Int("StatusCode", resp.StatusCode))
+		msg.Nack()
+		return
```
response body also needs to be closed in this case.

nachocano replied:
done, thanks!

nachocano commented on cmd/pubsub/publisher/main.go:
```diff
@@ -80,13 +87,21 @@ func main() {
-		logger.Fatal("failed to start publisher: ", zap.Error(err))
+	logger.Info("Starting publisher", zap.Any("publisher", publisher))
+	if err := publisher.Start(ctx); err != nil {
+		logger.Fatal("Failed to start publisher", zap.Error(err))
```
Done

nachocano commented:
Pushed to see if tests pass. Will debug later tonight if they do not...

nachocano commented:
/test pull-google-knative-gcp-unit-tests

yolocs commented on pkg/pubsub/adapter/converters/pubsub_pull.go:
```diff
@@ -0,0 +1,64 @@
+	. "github.com/google/knative-gcp/pkg/pubsub/adapter/context"
+)
+
+func convertPubSubPull(ctx context.Context, msg *pubsub.Message) (*cev2.Event, error) {
```
This is only used for pubsub channel, right?

nachocano replied:
this is actually used for PullSubscriptions. I decided to leave the logic there. We had some E2E tests for it.
But we have our Pullsubscription resources as internal, so users should actually be using our four sources, and not this stuff....

As we don't do any opinionated conversions for the channel, the channel just uses:

`binding.ToEvent(ctx, cepubsub.NewMessage(msg))`

yolocs commented on pkg/pubsub/adapter/converters/pubsub_pull.go:
```diff
@@ -0,0 +1,64 @@
+	}
+
+	event.SetSource(v1beta1.CloudPubSubSourceEventSource(project, topic))
+	event.SetType(v1beta1.CloudPubSubSourcePublish)
```
We can address this separately but I think this should be unset because we don't control the data in this case. Thoughts?

nachocano replied:
this is for an internal opinionated Pullsubscription not the Channel. 
I think is fine to leave it...

nachocano replied:
unless you think otherwise

nachocano commented on pkg/pubsub/adapter/converters/build_test.go:
```diff
@@ -18,122 +18,117 @@ package converters
-				"attribute1": "value1",
-				"attribute2": "value2",
-			}, buildID, buildStatus)
+		wantEventFn: func() *cev2.Event {
```
nope, we are not blindly promoting attributes to extensions any longer for any of our sources. The only one I left it as is was the PullSubscription, which is technically internal and we don't advertise it as a source. 

This is part of the work @yolocs has been doing on the schema definitions for our sources... More changes are needed after this refactor

nachocano replied:
and we should properly document in the release notes all those changes

ian-mi replied:
Should this be deleted then? Or can you switch this back to using cmp.Diff.

nachocano replied:
Done!

nachocano commented on pkg/pubsub/adapter/converters/pubsub.go:
```diff
@@ -20,71 +20,46 @@ import (
-	// Set the schema if it comes as an attribute.
-	if val, ok := msg.Attributes["schema"]; ok {
-		delete(msg.Attributes, "schema")
-		event.SetDataSchema(val)
```
this was again an opinionated implementation we did long time back. We were saying if they send us schema as an attribute, then we use it... This if actually never went through, unless some user send a pubsub message with an attribute named schema... 
We are not doing this anymore, we decided to always put the PubSub Push message format in the payload, which a well-defined schema

nachocano commented on pkg/pubsub/adapter/converters/pubsub_test.go:
```diff
@@ -23,126 +23,54 @@ import (
-		wantEventFn func() *cloudevents.Event
+		message     *pubsub.Message
+		wantEventFn func() *cev2.Event
 		wantErr     bool
```
my bad, it's because of the attributes change. Added an extra check for invalid context. Thanks!

nachocano commented on pkg/pubsub/adapter/converters/scheduler.go:
```diff
@@ -20,65 +20,41 @@ import (
-	event.Data = msg.Data
-	event.DataEncoded = true
-	// Attributes are extensions.
-	if msg.Attributes != nil && len(msg.Attributes) > 0 {
```
we are not... we do not blindly promote attributes to extensions. We might set some attributes. In the above case, we use some for the CE source and CE subject and then delete them. Those values were populated from our controller, so it's fine to remove... 

Why do you think is necessary to delete them?

ian-mi replied:
I don't, I'm questioning whether that code can be removed.

nachocano replied:
got it, yes, the code can be removed

nachocano replied:
I did remove it?

nachocano commented on pkg/reconciler/intevents/reconciler.go:
```diff
@@ -30,7 +30,7 @@ import (
 	"k8s.io/apimachinery/pkg/api/equality"
 	apierrs "k8s.io/apimachinery/pkg/api/errors"
-	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
```
yeah, some goland IDE thing... let me add it back

nachocano replied:
added

nachocano commented on pkg/reconciler/intevents/pullsubscription/resources/receive_adapter_test.go:
```diff
@@ -205,7 +207,7 @@ func TestMakeFullReceiveAdapter(t *testing.T) {
 			},
 			Topic:       "topic",
-			AdapterType: "adapter-type",
+			AdapterType: string(converters.PubSubPull),
```
Done! thanks!

nachocano commented on pkg/tracing/attributes.go:
```diff
@@ -30,6 +30,10 @@ var (
 }
+
+func SourceDestination(s types.NamespacedName) string {
+	return fmt.Sprintf("source:%s.%s", s.Name, s.Namespace)
```
another good point. Done, thanks Ian!

ian-mi commented on go.mod:
```diff
@@ -7,7 +7,7 @@ require (
 	cloud.google.com/go/pubsub v1.3.2-0.20200506222144-2c46308f8465
 	cloud.google.com/go/storage v1.8.0
-	github.com/cloudevents/sdk-go v1.2.0
+	github.com/cloudevents/sdk-go v1.0.0
```
Was this intentional?

nachocano commented on pkg/pubsub/adapter/adapter.go:
```diff
@@ -18,253 +18,252 @@ package adapter
-	if a.SendMode == "" {
-		a.SendMode = converters.DefaultSendMode
-	}
+// Stop stops the adapter.
+func (a *Adapter) Stop() {
+	a.cancel()
+}
```
will be once I get the UTs working. 
Mind if I leave it?

nachocano commented on pkg/pubsub/adapter/adapter.go:
```diff
@@ -18,253 +18,252 @@ package adapter
-		cloudevents.WithTarget(target),
+func (a *Adapter) startSpan(ctx context.Context, event *cev2.Event) (context.Context, *trace.Span) {
+	spanName := tracing.SourceDestination(a.namespacedName)
+	// This receive adapter code is used both for Sources and Channels.
+	// An ugly way to identify whether it was created from a Channel is to look at the resourceGroup.
```
yeah, this is old stuff but mainly due to metrics. At the time we wanted to know the resourceGroup of a source (e.g., cloudpubsubsources, cloudauditlogssources, etc)... so that we can report metrics from particular sources types and not just a generic `Source`, that is why this ugly resourceGroup stuff...
If you don't mind, I'd rather leave it as is for now, and then once we do a pass on metrics, we can revisit.. Sounds good?

nachocano commented on cmd/pubsub/receive_adapter/main.go:
```diff
@@ -72,26 +144,62 @@ func main() {
+
+	args := &AdapterArgs{
+		TopicID:        env.Topic,
+		ConverterType:  converters.ConverterType(env.AdapterType),
```
if it's not an adapter registered (e.g., CAL, PubSub, Storage, Scheduler, or PubSubPull), we just use the default conversion, which is basically doing:

`binding.ToEvent(ctx, cepubsub.NewMessage(msg))`

nachocano commented:
Darn, sorry guys! I updated deps to try to get rid of cloudevents sdk v1.. the import is still there, not sure why...
But now there is a breaking change in eventing. The KRShaped stuff... Will fix that and continue with the review comments...

My apologies because this is actually turning into a much bigger issue

nachocano commented:
The genreconciler changed and uses KRShaped. Implemented KRShaped in all the types we use in reconcilers. Mainly v1beta1 types and v1alpha1 of BrokerCell.
Removed all the old generated reconciler code. 

It seems that a trigger_test is failing. 
Pinged @liu-cong as he might now better...

nachocano commented on pkg/pubsub/adapter/adapter.go:
```diff
@@ -246,11 +246,17 @@ func (a *Adapter) sendMsg(ctx context.Context, address string, msg binding.Messa
+	// This receive adapter code is used both for Sources and Channels.
+	// An ugly way to identify whether it was created from a Channel is to look at the resourceGroup.
+	if a.resourceGroup == messaging.ChannelsResource.String() {
+		spanName = tracing.ChannelDestination(a.namespacedName)
```
Done

nachocano commented on pkg/pubsub/adapter/adapter.go:
```diff
@@ -46,87 +66,66 @@ type Adapter struct {
+
 	go func() {
-		errCh <- a.subscription.Receive(ctx, a.receive)
+		done(a.subscription.Receive(ctx, a.receive))
```
I was following what was done in fanout https://github.com/google/knative-gcp/blob/master/pkg/broker/handler/fanout.go#L178
I had a more convoluted version before and Chen commented on..

You mean just do:
`return a.subscription.Receive(ctx, a.receive)` ?

I think it will be easier as is for writing the UTs... Unless you have a strong objection, I'd rather leave it for now... WDYT

ian-mi replied:
Yes, async callbacks are often an antipattern in go. A synchronous api can always be easily converted to an async callback by the caller using a goroutine.

nachocano replied:
SG, done, thanks for the feeback

liu-cong commented:
The root cause of the unit test failures is that the new genreconciler now unconditionally calls `PreProcessReconcile` which sets the init conditions if not set. Previously the conditions are only set by our own reconcilers.

nachocano commented:
> The root cause of the unit test failures is that the new genreconciler now unconditionally calls `PreProcessReconcile` which sets the init conditions if not set. Previously the conditions are only set by our own reconcilers.

Thanks @liu-cong !!!

nachocano commented:
/test pull-google-knative-gcp-unit-tests

nachocano commented:
/test pull-google-knative-gcp-wi-tests

nachocano commented:
/retest pull-google-knative-gcp-unit-tests

nachocano commented:
/test pull-google-knative-gcp-unit-tests

nachocano commented:
/test pull-google-knative-gcp-unit-tests

nachocano commented:
/retest

On Wed, Jun 24, 2020 at 12:43 AM Knative test reporter robot <
notifications@github.com> wrote:

> The following jobs failed:
> Test name Triggers Retries
> pull-google-knative-gcp-wi-tests pull-google-knative-gcp-wi-tests
> <https://prow.knative.dev/view/gcs/knative-prow/pr-logs/pull/google_knative-gcp/1296/pull-google-knative-gcp-wi-tests/1275660442689081345>
> pull-google-knative-gcp-wi-tests
> <https://prow.knative.dev/view/gcs/knative-prow/pr-logs/pull/google_knative-gcp/1296/pull-google-knative-gcp-wi-tests/1275668746555363328>
> pull-google-knative-gcp-wi-tests
> <https://prow.knative.dev/view/gcs/knative-prow/pr-logs/pull/google_knative-gcp/1296/pull-google-knative-gcp-wi-tests/1275677807183138816>
> pull-google-knative-gcp-wi-tests
> <https://prow.knative.dev/view/gcs/knative-prow/pr-logs/pull/google_knative-gcp/1296/pull-google-knative-gcp-wi-tests/1275687369739079685>
> 3/3
>
> Job pull-google-knative-gcp-wi-tests expended all 3 retries without
> success.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/knative-gcp/pull/1296#issuecomment-648653942>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABD65DCI7LHFD3EUXO5YTPTRYGVDTANCNFSM4OAXLQIQ>
> .
>

nachocano commented:
/test pull-google-knative-gcp-wi-tests

nachocano commented:
/test pull-google-knative-gcp-wi-tests

nachocano commented:
pull-google-knative-gcp-wi-tests is failing pretty much all the time. I'm thinking of disabling them if folks are having the same issues in other PRs

nachocano commented:
Seems WI tests are not required to merge...
Waiting until they run one more time, and the UTs as well (there was a flake on the configMap assertion). 
I think we can get this in to unblock people and I can focus on the follow ups... @yolocs @ian-mi

nachocano commented:
/test pull-google-knative-gcp-wi-tests

ian-mi commented on pkg/pubsub/adapter/converters/storage_test.go:
```diff
@@ -19,171 +19,151 @@ package converters
-		sendMode    ModeType
-		wantEventFn func() *cloudevents.Event
+		message     *pubsub.Message
+		wantEventFn func() *cev2.Event
```
This is not being used either

nachocano replied:
Mind if we do in a follow up?

nachocano replied:
done

ian-mi commented on pkg/pubsub/adapter/stats_reporter_test.go:
```diff
@@ -27,17 +27,12 @@ import (
 	}
 
-	r := NewStatsReporter()
+	r, _ := NewStatsReporter("testobject", "testns", "testresourcegroup")
```
err here should be fatal

nachocano replied:
ack, let me do in a follow up

nachocano replied:
done

nachocano commented on pkg/pubsub/adapter/converters/storage_test.go:
```diff
@@ -19,171 +19,151 @@ package converters
-		sendMode: Binary,
-		wantEventFn: func() *cloudevents.Event {
-			return storageCloudEvent(map[string]string{})
+		wantEventFn: func() *cev2.Event {
```
it is used here?

ian-mi replied:
it's set here but I don't see where it's used

nachocano replied:
actually this should be removed, you are right... 
doing it in a follow up

nachocano commented:
/test pull-google-knative-gcp-unit-tests

nachocano commented:
Agreed with @ian-mi and @yolocs to get this in and I'll address the unresolved comments in the follow up. 
We need to unblock @yolocs work and allow @capri-xiyue to get new numbers on receive adapter resource requests and limits

nachocano commented:
@ian-mi addressed the tests comments so that I don't forget. @yolocs will have to do changes either way and I'll also improve stuff with the follow up

nachocano commented:
Need one more lgtm

ian-mi commented on pkg/pubsub/adapter/converters/scheduler.go:
```diff
@@ -20,65 +20,41 @@ import (
+	jobName, ok := msg.Attributes[v1beta1.CloudSchedulerSourceJobName]
 	if ok {
-		delete(msg.Attributes, v1alpha1.CloudSchedulerSourceJobName)
+		delete(msg.Attributes, v1beta1.CloudSchedulerSourceJobName)
```
I think my comment below was ambiguous, I was thinking that the deletion of these attributes is no longer necessary.

nachocano replied:
ohh I think they are, because with them we populate ce-source and ce-subject...

nachocano replied:
darn, I get your point... no need to delete them argggg

nachocano replied:
the arguments are necessary, no need to delete, let me remove that

nachocano replied:
Done, thanks for the catch

yolocs commented:
/lgtm

ian-mi commented:
/lgtm
Thanks for the changes!
